### ISSUE TITLE
Improving the Visual Studio test adapter

### ISSUE THREAD

**Issue by seesharper** [fixie/fixie#85]
One thing I have noticed using this really nice library is that the Visual Studio test adapter could need some TLC :) Most notably it does not allow you to run the tests from code using the standard shortcuts and nor does it navigate to the test method when double clicked in the test explorer. 

I have done some research on this and I think that the safest bet here is Mono.Cecil as it provides a stable and proven library for reading (and writing) .Net assemblies. Not only can it read the contents of an assembly, but it can also read the symbols which is really what we are looking for here. 

What needs to be done is to provide the source file location along with the line number to the TestCase class that is used during test dicovery. I did a simple test for this and once the source file and line number gets specified, the navigation and running tests from code works just great.

I have created a class for this that wraps Mono.Cecil to read out this information. It can also find the source location for async methods. Even xUnit does not handle async methods with regards to navigation and running tests from code. 

If you guys are willing to take a dependency upon Mono.Cecil, I will gladly create a PR for this as I feel that it would provide great value to an already great library.

Anyways, here is the code that reads out the source location.

```
namespace SourceLocationSpike
{
    using System.Collections.Concurrent;
    using System.Linq;
    using Mono.Cecil;
    using Mono.Cecil.Cil;
    using Mono.Cecil.Rocks;

    /// <summary>
    /// Represents the source location of a test method.
    /// </summary>
    public class SourceLocation
    {        
        /// <summary>
        /// Gets the path of the source file that contains the test method.
        /// </summary>
        public string Path { get; internal set; }

        /// <summary>
        /// Gets the line number of the test method.
        /// </summary>
        public int LineNumber { get; internal set; }
    }

    /// <summary>
    /// Represents a class that is capable of providing the <see cref="SourceLocation"/>
    /// for a given test method.
    /// </summary>
    public interface ISourceLocationProvider
    {        
        /// <summary>
        /// Gets the <see cref="SourceLocation"/> for the given <paramref name="methodName"/>.
        /// </summary>
        /// <param name="assemblyFileName">The path of the assembly that contains the method.</param>
        /// <param name="className">The name of the class that contains the method.</param>
        /// <param name="methodName">The name of the method for which to return the <see cref="SourceLocation"/>.</param>
        /// <returns>The <see cref="SourceLocation"/> if found, otherwise, null.</returns>
        SourceLocation GetSourceLocation(string assemblyFileName, string className, string methodName);
    }

    /// <summary>
    /// A <see cref="ISourceLocationProvider"/> that uses Mono.Cecil to 
    /// read method symbol information. 
    /// </summary>
    public class SourceLocationProvider : ISourceLocationProvider
    {        
        private readonly ConcurrentDictionary<string, TypeDefinition[]> moduleTypeCache =
            new ConcurrentDictionary<string, TypeDefinition[]>();

        private readonly ConcurrentDictionary<string, TypeDefinition> typeCache =
            new ConcurrentDictionary<string, TypeDefinition>();

        /// <summary>
        /// Gets the <see cref="SourceLocation"/> for the given <paramref name="methodName"/>.
        /// </summary>
        /// <param name="assemblyFileName">The path of the assembly that contains the method.</param>
        /// <param name="className">The name of the class that contains the method.</param>
        /// <param name="methodName">The name of the method for which to return the <see cref="SourceLocation"/>.</param>
        /// <returns>The <see cref="SourceLocation"/> if found, otherwise, null.</returns>
        public SourceLocation GetSourceLocation(string assemblyFileName, string className, string methodName)
        {
            MethodDefinition testMethod = GetMethod(assemblyFileName, className, methodName);
            SequencePoint sequencePoint;
            CustomAttribute asyncStateMachineAttribute;
            if (TryGetAsyncStateMachineAttribute(testMethod, out asyncStateMachineAttribute))
            {
                var stateMachineMoveNextMethod = GetStateMachineMoveNextMethod(asyncStateMachineAttribute);
                sequencePoint = GetMethodSequencePoint(stateMachineMoveNextMethod.Body);
            }
            else
            {
                sequencePoint = GetMethodSequencePoint(testMethod.Body);
            }

            if (sequencePoint == null)
            {
                return null;
            }
            
            return new SourceLocation
                       {
                           Path = sequencePoint.Document.Url,
                           LineNumber = sequencePoint.StartLine - 1
                       };
        }

        private static MethodDefinition GetStateMachineMoveNextMethod(CustomAttribute asyncStateMachineAttribute)
        {
            var stateMachineType = (TypeDefinition)asyncStateMachineAttribute.ConstructorArguments[0].Value;
            var stateMachineMoveNextMethod = stateMachineType.GetMethods().First(m => m.Name == "MoveNext");
            return stateMachineMoveNextMethod;
        }

        private static bool TryGetAsyncStateMachineAttribute(MethodDefinition method, out CustomAttribute attribute)
        {
            attribute =
                method.CustomAttributes.FirstOrDefault(c => c.AttributeType.Name == "AsyncStateMachineAttribute");
            return attribute != null;
        }

        private static TypeDefinition[] GetTypes(string assemblyName)
        {
            ReaderParameters readerParameters = new ReaderParameters { ReadSymbols = true };
            ModuleDefinition module = ModuleDefinition.ReadModule(assemblyName, readerParameters);
            return module.GetTypes().ToArray();
        }

        private static SequencePoint GetMethodSequencePoint(MethodBody body)
        {
            foreach (var instruction in body.Instructions)
            {
                if (instruction.SequencePoint != null && instruction.SequencePoint.StartLine != 16707566)
                {
                    return instruction.SequencePoint;
                }
            }

            return null;
        }

        private MethodDefinition GetMethod(string assemblyFileName, string className, string methodName)
        {
            var type = typeCache.GetOrAdd(className, cn => GetType(assemblyFileName, cn));                                    
            return type.GetMethods().First(m => m.Name == methodName);
        }

        private TypeDefinition GetType(string assemblyFileName, string className)
        {
            TypeDefinition[] types = moduleTypeCache.GetOrAdd(assemblyFileName, GetTypes);
            return types.First(t => t.FullName == className);    
        }               
    }
}
```

Best regards 
Bernhard Richter

**Comment by plioi** [comment]
Thanks a lot for your research.  Setting the line numbers has been on the to-do list, but I hadn't put as much research into it yet as you have.  Let's discuss the plan a little more before starting on a pull request:

I see that NUnit and xUnit both use a different dependency, something called DiaSession which lives inside the Microsoft.VisualStudio.TestPlatform.ObjectModel assembly that we're already referencing for the rest of the Visual Studio runner functionality.  Pretending for a moment that DiaSession works perfectly, it would be preferable as one less dependency.

You can see those existing usages inside NUnit's Visual Studio Adapter https://github.com/nunit/nunit-vs-adapter/blob/master/src/NUnitTestAdapter/TestConverter.cs and inside xUnit itself https://github.com/xamarin/xunit/blob/master/src/xunit.runner.utility/Utility/DiaSession.cs (xUnit's Visual Studio Adapter makes *use* of the calculated line numbers, but they're discovered within xUnit itself).

If Mono.Cecil is getting line number information from a compiled assembly, I'd be very suspicious of the accuracy, especially when things happen to be compiled in Release mode, in which whole methods can disappear between the original source and the compiled assembly.  For instance, the assertion library called Shouldly does something very similar with untrustworthy line numbers, and the results get quite buggy when compiler optimizations are used.  If I'm wrong about what Cecil would be doing to get accurate line numbers, please let me know.

The sample code above appears to work with the method name as a string.  What if the method is overloaded?

Lastly, you called out a specific problem with xUnit for async methods.  Does the same problem happen with NUnit?

**Comment by seesharper** [comment]
Hi again and thanks a lot for your response.

First of all the reason that the method name is a string is that VsTestDiscoverer seems to work with methods as strings. If we had the actual MethodInfo it would not only simplify the signature of the GetSourceLocation method, but it would also handle method overloading although I am unsure how common it is to have overloaded test methods.

So, when it comes to Mono.Cecil I have just tried this using a release build and given that we have the pdb available it still works as expected. After all, who runs tests without the pdb? :)

When it comes to accuracy, I think the sequence points provides the closest thing possible. After all, this is the exact same source of information that the stack trace  use in the event of an exception. We don't have to worry about method inlining here since we are talking about public method that don't get inlined. Optimization within the method itself it not so important here since we are basically looking for a valid sequence point and the beginning of the method. If we were to look for a particular statement it would have been a different story. 

I have also looked into xUnit and quite frankly my impression was that the code that handles this seemed a bit complex. In xUnit, navigation to an async test method does not work under certain conditions. It depends on the body of the test  method. I'll try to come up with an example of such a test.

I understand that you are reluctant to take a dependency on a third party assembly and I also pick those dependencies with extreme care. This, however is a complex area and a "do-it-yourself" approach might not be the best solution :)

One option here is to use ILMerge to "embed" Mono.Cecil in the test adapter assembly. ILMerge kan internalize Cecil so that it is not visible to the outside world. 

Bottom line here is that if we can provide a solution that works 90% of the time with 10% of the effort, we have still provided  an immense value to the users of Fixie. 

Best regards 
Bernhard Richter

**Comment by seesharper** [comment]
I have created a fork at https://github.com/seesharper/fixie that uses the Cecil approch. We are currently testing this out at my company. So far it seem to work as expected. 
Check it out and if you like I will send you a PR.

PS: I had to modify the a couple of tests to use a locale specific decimal separator. In Norway we use "," and not "." :)

**Comment by seesharper** [comment]
What is the status of this issue. Are you guys working on the Visual Studio test adapter or would you be interested in a PR that uses the Cecil approach?

**Comment by plioi** [comment]
I'm planning on implementing a number of test runner enhancements this week, including handling these line numbers.

**Comment by seesharper** [comment]
Great news. That would be an awesome addition to this fabulous framework.

**Comment by plioi** [comment]
Just to catch you up on my progress: I ran into a few extra complications with this solution (oddball cases like nested classes, inherited tests, overloads, so-called "hidden" lines...), but I've found a good way to test-drive the details and have gotten past the obstacles.  I expect to have a working solution in a few days.

**Comment by seesharper** [comment]
That is great news, Patrick. I'm looking forward to this and will replace
xUnit for Fixie once this has been resolved :)

**Comment by plioi** [comment]
Closed by #89.

**Comment by seesharper** [comment]
I see that you went with the Cecil approach I suggested after all :)
Great work, Patrick. Thanks a lot for your help.

Best regards
Bernhard Richter

**Comment by plioi** [comment]
Yeah.  I tried at first to mimic xUnit and NUnit's use of a class called DiaSession, but they had to jump through so many hoops that it was getting too complicated.
